### PR TITLE
SSD Inference

### PR DESCRIPTION
**Pull request by ArutyunovG**
This pull request adds SSD inference support in Caffe2.

### CODE REVIEW

**Comment by ArutyunovG** [comment]
Reference #13896.

**Comment by ArutyunovG** [comment]
@BIT-silence @orionr 

Hi,

We have completed the task of conversion of SSD models from Caffe and TF and the pull request of Caffe2 changes is ready.

To simplify giving this pull request a try and see how the changed stuff works, we have zipped five different models: two from Caffe and three from TF OD API.

This is [the link](https://drive.google.com/file/d/1yAGB18dvx2Ell4WO-VQEevbnqyOEWJSf/view?usp=sharing) to the archive of all the models in three formats: 1) original; 2) DLDT's IR; 3) Caffe2.

This pull request will simplify our working process a lot, since we now can use only only one framework for inference of 
1) SSD models, trained in 
1.1) Caffe;
1.2) TF OD API, 
as well as 
2) Faster R-CNN models, trained in 
2.1) Caffe and 
2.2) Caffe2.

Faster/Mask R-CNNs from TF OD API do not convert at the moment, since they are based on CropAndResize opeartion instead of ROIPool/ROIAlign.

Actually, the C++ detection API on top of Caffe2 for SSDs is completely identical to the one with Faster/Mask R-CNNs in their proto formats.

The conversion is performed in the following way:

Original Framework => [DLDT] => IR => [Our code] => Caffe2. 

The conversion code is too premature to open source at the moment though, since we only started its development process.

Thanks, 
Grigory

**Comment by Georg-W** [comment]
Hi @ArutyunovG ,
thank you for the contribution, I am very interested in running the tensorflow models on caffe2 so this is really helping me.
Is there any way to get some intuition about the conversion code? Or sample scripts that you used for testing the ssdlite mobilenet v2 with your version of caffe2?

Thanks,
Georg

**Comment by orionr** [comment]
@newstzpz, would you be interested in reviewing as well?

**Comment by ArutyunovG** [comment]
Hi @Georg-W,

Thanks for your interest, nice to see that our work is helpful for others.

The source code for conversion is in too premature state at the moment to be open sourced. We started the work on it just two weeks ago. What we have at the moment is enough though, to be sure that pull requests like this one make sense. But it is too early to say, that we have done enough to open.

When we get some state, which will seem to be reasonable, I'm planning to open source it, so other people can use, test and contribute new converted models.

As an inspiration we use [Model Optimizer](https://github.com/opencv/dldt/tree/2018/model-optimizer) from DLDT. It is an amazing conversion toolset created by Intel developers. 

The testing is performed at the moment by some of our internal tools, which dumps intermediate blobs in DLDT and Caffe2 to text files and we just compare those blobs with python.

Please, write me an e-mail to ArutyunovG@yandex.ru, so I can e-mail back, when the source will be opened.

Grigory

**Comment by ArutyunovG** [comment]
We are close to add support for faster r-cnn meta-architecture from TF OD API.
Changes there overlap with those in this PR, so I'm closing this as deprecated in order to save reviewers' time. 
We'll open another PR for both SSD and FRCNN meta-architectures inference.

**Comment by ArutyunovG** [comment]
Guys, just want to say, that we are still alive here, in case someone is interested and recieves this as e-mail notification :)
This is the [current state](https://github.com/ArutyunovG/pytorch/commit/8c91475c6f294e670c3e64ceda0c389af49cd234) of the work. 
The most difficult part was to make TF models batchable and reshapable. They are not reshapable in frozen state originally.
It is over and was solved mostly by our amazing colleagues from other team.
As you can see, we now have support for both SSD models, as well as CropAndResize. All of it allows to run both SSDs, as well as FRCNNs from TF OD API in Caffe2. 
We already started to use Caffe2 as inference engine for NVIDIA GPUs in our solutions.
As for PR we need to test anything properly and write missing tests for changes.
We are alive with this, best :)
